kani-compiler/src/kani_middle/points_to/points_to_graph.rs

```diff
@@ -0,0 +1,216 @@
+                }
+            }
+        }
+        node_set
+    }
+
+    /// Dump the graph into a file using the graphviz format for later visualization.
+    pub fn dump(&self, file_path: &str) {
+        let mut nodes: Vec<String> =
+            self.edges.keys().map(|from| format!("\t\"{:?}\"", from)).collect();
+        nodes.sort();
+        let nodes_str = nodes.join("\n");
+
+        let mut edges: Vec<String> = self
+            .edges
+            .iter()
+            .flat_map(|(from, to)| {
+                let from = format!("\"{:?}\"", from);
+                to.iter().map(move |to| {
+                    let to = format!("\"{:?}\"", to);
+                    format!("\t{} -> {}", from.clone(), to)
+                })
+            })
+            .collect();
+        edges.sort();
+        let edges_str = edges.join("\n");
+
+        std::fs::write(file_path, format!("digraph {{\n{}\n{}\n}}", nodes_str, edges_str)).unwrap();
+    }
+
+    /// Find a transitive closure of the graph starting from a set of given locations; this also
+    /// includes statics.
+    pub fn transitive_closure(&self, targets: HashSet<GlobalMemLoc<'tcx>>) -> PointsToGraph<'tcx> {
+        let mut result = PointsToGraph::empty();
+        // Working queue.
+        let mut queue = VecDeque::from_iter(targets);
+        // Add all statics, as they can be accessed at any point.
+        let statics = self.edges.keys().filter(|node| matches!(node, GlobalMemLoc::Global(_)));
+        queue.extend(statics);
+        // Add all entries.
+        while !queue.is_empty() {
+            let next_target = queue.pop_front().unwrap();
+            result.edges.entry(next_target).or_insert_with(|| {
+                let outgoing_edges =
+                    self.edges.get(&next_target).cloned().unwrap_or(HashSet::new());
+                queue.extend(outgoing_edges.iter());
+                outgoing_edges.clone()
+            });
+        }
+        result
+    }
+
+    /// Retrieve all places to which a given place is pointing to.
+    pub fn pointees_of(&self, target: &GlobalMemLoc<'tcx>) -> HashSet<GlobalMemLoc<'tcx>> {
+        self.edges.get(&target).unwrap_or(&HashSet::new()).clone()
+    }
+
+    // Merge the other graph into self, consuming it.
+    pub fn consume(&mut self, other: PointsToGraph<'tcx>) {
+        for (from, to) in other.edges {
+            let existing_to = self.edges.entry(from).or_default();
+            existing_to.extend(to);
+        }
+    }
+}
+
+/// Since we are performing the analysis using a dataflow, we need to implement a proper monotonous
+/// join operation. In our case, this is a simple union of two graphs. This "lattice" is finite,
+/// because in the worst case all places will alias to all places, in which case the join will be a
+/// no-op.
+impl<'tcx> JoinSemiLattice for PointsToGraph<'tcx> {
+    fn join(&mut self, other: &Self) -> bool {
```

Comment:
Isn't this supposed to be calling the function above? Why are there two different ways to join two PointsToGraph? Also, one seems to be removing redundant edges, while the other doesn't.

If you want to remove redundant edges, you might was to use a `HashSet` to represent the value in edges.

Reply:
`consume` is supposed to be a faster alternative to `join`, since sometimes we can safely pass the graph by value and we don't really care about whether anything gets changed or not. The signature of `join` is mandated by the `JoinSemiLattice` implementation, which we must implement to use the dataflow solver. The value of `edges` is already a `HashSet`, so in both cases redundant edges are removes.

Reply:
Is the performance difference significant? If you are already using HashSet, why do you need to compute the difference? Can you check the size before and after the update?

Reply:
There used to be some performance difference, but I believe it has been mitigated by other improvements. Since this is the case, I agree that having only one "join" method is cleaner. I made the necessary changes